### PR TITLE
TMDM-14523 Password is in clear

### PR DESCRIPTION
**Pull request by sbliu**
**What is the current behavior?** (You should also link to an open issue here)
https://jira.talendforge.org/browse/TMDM-14523
Password still show in source editor for workflow & workflowcontext service jndi name, or for workflowtrigger & workflowcontexttrigger plugin for process.

**What is the new behavior?**
Hide password in source editor. But allow user edit all parameters in source editor, if changed password contains * , all other changed parameters exception password will be saved.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 1]
```diff
@@ -14,8 +14,18 @@
 
 public class ExtensibleEditorContent {
 
+    private static final String TAG_PASSWORD_GEGIN = "<password>";
```
Comment (pbabin) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:17]: GEGIN ?

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 2]
```diff
@@ -14,8 +14,18 @@
...
+
+    private static final String TAG_PASSWORD_END = "</password>";
+
+    private static final String MASKCODE_CHAR = "*";
+
+    private static final String MASKCODE = "*******";
+
     protected String content = "";//$NON-NLS-1$
 
+    protected String maskedContent = "";//$NON-NLS-1$
```
Comment (pbabin) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:27]: Can't both above variables be private?

Comment (pbabin) [reply main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:27]: Also they are initialized , but later on in the code reinitilized also. There could be better code there.

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 3]
```diff
@@ -26,11 +36,66 @@ public String getContent() {
...
+    private String hidePassword(final String content) {
+        String[] splitbyPasswordTag = splitbyPasswordTag(content);
+        if (splitbyPasswordTag != null) {
+            return splitbyPasswordTag[0] + (TAG_PASSWORD_GEGIN + MASKCODE + TAG_PASSWORD_END) + splitbyPasswordTag[2];
+        }
+
+        return content;
+    }
+
+    private boolean isPasswordHiden(final String content) {
```
Comment (hbhong) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:71]: Hiden -> Hidden

File: test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                + "  <password>talend</password>\r\n" + "  <useBuiltInVariable>false</useBuiltInVariable>\r\n"
+                + "  <defaultDataModel/>\r\n" + "</parameters>";
+
+        String[] expected = new String[] {
+                "<parameters>\r\n" + "  <processId>Product_Product</processId>\r\n" + "  <processVersion>1.0</processVersion>\r\n"
+                        + "  <username>admin</username>\r\n  ",
+                "<password>talend</password>",
+                "\r\n" + "  <useBuiltInVariable>false</useBuiltInVariable>\r\n" + "  <defaultDataModel/>\r\n" + "</parameters>" };
+
+        try {
```
Comment (pbabin) [inline test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java:38]: Are all cases covered here base on the code added?
password missing, password empty, password already masked, etc... ?

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 5]
```diff
@@ -26,11 +36,66 @@ public String getContent() {
...
             this.content = "";//$NON-NLS-1$
-        else
-            this.content = content;
+            this.maskedContent = "";
+        } else {
+            if (isPasswordHiden(newContent)) {
+                if (contentChanged(newContent)) {
+                    String[] splitNewContent = splitbyPasswordTag(newContent);
+                    String[] splitContent = splitbyPasswordTag(content);
+                    this.content = splitNewContent[0] + splitContent[1] + splitNewContent[2];
```
Comment (hbhong) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:52]: splitContent  or splitNewContent may be null, should has a check if it is null

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 6]
```diff
@@ -26,11 +36,66 @@ public String getContent() {
...
+
+        if(splitbyPassword_newcontent != null && splitbyPassword_maskcontent != null) {
+            return !(splitbyPassword_newcontent[0].equals(splitbyPassword_maskcontent[0])
+                    && splitbyPassword_newcontent[2].equals(splitbyPassword_maskcontent[2]));
+        }
+        
+        return true;
+    }
+
+    private String[] splitbyPasswordTag(String _content) {
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:92]: `splitByPasswordTag`

File: test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java (hunk 4 above)
Comment (sbliu) [reply test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java:38]: will add two case password missing , password empty. case password already mask same as password not masked.

File: test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+
+public class ExtensibleEditorContentTest {
+
+    @Test
+    public void testSplitbyPasswordTag() {
+        // check cases:common password, masked password
+        String[] checkPasswordPart = { "<password>talend</password>", "<password>******</password>",
```
Comment (lpw82gf) [inline test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java:28]: Why define 2 same string array?

Comment (lpw82gf) [reply test/plugins/org.talend.mdm.workbench.test/src/main/java/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContentTest.java:28]: How about define `String[] passwordPart` for both `check` and `expected` usage.

File: main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java [hunk 8]
```diff
@@ -26,11 +36,78 @@ public String getContent() {
...
+                    this.maskedContent = hidePassword(content);
+                }
+            } else {
+                this.content = newContent;
+                this.maskedContent = hidePassword(content);
+            }
+        }
+    }
+
+    private String hidePassword(final String content) {
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:69]: The following `private` methods seems util method, can we move to a util class.

Comment (lpw82gf) [reply main/plugins/org.talend.mdm.workbench/src/com/amalto/workbench/widgets/xmleditor/ExtensibleEditorContent.java:69]: Better to add JUnit for all the `private` methods.

**Pull request merged**